libzeth/circuits/mimc/mimc_round_gadget.tcc

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef __ZETH_CIRCUITS_MIMC_ROUND_GADGET_TCC__
+#define __ZETH_CIRCUITS_MIMC_ROUND_GADGET_TCC__
+
+#include "libzeth/circuits/mimc/mimc_round_gadget.hpp"
+
+namespace libzeth
+{
+
+template<typename FieldT, size_t Exponent>
+MiMC_round_gadget<FieldT, Exponent>::MiMC_round_gadget(
+    libsnark::protoboard<FieldT> &pb,
+    const libsnark::pb_variable<FieldT> &x,
+    const libsnark::pb_variable<FieldT> &k,
+    const FieldT &c,
+    libsnark::pb_variable<FieldT> &result,
+    const bool add_k_to_result,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _x(x)
+    , _k(k)
+    , _c(c)
+    , _result(result)
+    , _add_k_to_result(add_k_to_result)
+{
+    // Each condition requires an intermediate variable, except the final one,
+    // which uses _result (and optionally _k).
+    _exponents.resize(NUM_CONDITIONS - 1);
+}
+
+template<typename FieldT, size_t Exponent>
+void MiMC_round_gadget<FieldT, Exponent>::generate_r1cs_constraints()
+{
+    // Mask to capture the last bit
+    constexpr size_t mask = 1 << (EXPONENT_NUM_BITS - 1);
```

Comment:
`Exponent` is a constant `size_t` (and in fac compile-time constant) so there should be no need to worry about endianness here.

Reply:
My point was mostly around the meaning of "last bit" in your comment: `// Mask to capture the last bit`. Seems like it is used to capture the MSB (if I follow correctly). If so, it may be worth referring to bits as MSB/LSB instead of "first and last"

Reply:
Ah, OK.  I see the point. I'll reword the comment.

Reply:
Reworded the comment.